networkx/algorithms/tests/test_d_separation.py

```diff
@@ -0,0 +1,79 @@
+from itertools import combinations
+import pytest
```

Comment:
@dschult @arunwise I noticed we're using `pytest` nowadays. Any chance that we could start having new tests use the `def test_...` style instead of the old unittest class-style? Or are you planning on keeping the old unittest style, @dschult? I just think it the isolated functions would make isolating the tests more easy, and prevent accidental future cross-talk between test examples because of human error. (We also wouldn't have to deal with boilerplate repeat!) In this paradigm, we would isolate each of the graphs (path, fork, collider, NB and asia) into their own test examples, rather than have them all encapsulated in a single class.

Reply:
We are definitely using pytest style now. But pytest allows both Class and Function interfaces. The Function interface is often easier to read, but the class interface allows a single setup method to create the graphs and then each of the test methods use those graphs. In that sense this implementation is probably good style because the setup method creates multiple graphs and the first two test methods use those multiple graphs.

So, the class style for tests is fine. And there is a choice whether to split out all tests for one graph into its own function vs make a class for all graphs with fewer test methods. What do you think @ericmjl ? Would it be more clear with a function for each graph, testing all aspects for that graph? Or is it better to have a method for each type of test and the graphs all get created in the setup method?  Either is perfectly good pytest syntax. pytest allows both functions or classes.

Reply:
@dschult I'll eventually defer to your preference (being the lead maintainer) :smile:, but in case my opinion counts for anything :laugh:, I definitely lean towards fixtures + functions rather than classes.

I know it's not good to prematurely optimize, but I think in this case, having each of the four graphs available as a fixture to be used in other places may be helpful as more graph-based causality algorithms get implemented.

Also, I think it's possibly beneficial to keep test examples as isolated as possible from one another, because of the reasons I argued for above. 

By separating out the tests, we make reading the output of any errors and isolating the exact test a bit more easy. Also, doing so makes possible adding docstrings to each of the tests (highly recommended practice!) which explain _why_ a test exists and what to expect from it, in plain English.

Of course, this isn't one of those hills I'm willing to die on, so I don't mind if this suggestion isn't accepted - no hard feelings! That said, I did also take the liberty of reworking some of the tests in a [gist](https://gist.github.com/ericmjl/43d14da2d2cde5e632e78cdcd816d5c0), to show how the test suite could be organized, with all of the goodies I mentioned above: fixtures, docstrings, etc. It should also lower the barrier for @arunwise to change the test suite style, if he also chooses to do so.

Reply:
@ericmjl Thanks for the [gist](https://gist.github.com/ericmjl/43d14da2d2cde5e632e78cdcd816d5c0)

@ericmjl @dschult I personally use the pytest fixture approach when writing tests. However, I noticed that almost all other modules used unittest approach. So, to keep things uniform I wrote the tests in unittest style. I don't have a strong opinion in this particular case, but generally I lean towards fixtures for reasons mentioned by @ericmjl

Reply:
This all sounds good. pytest fixtures are good.  And FWIW, any setup function put in a class can be created as a function and then called from the test_function as well. :}

Reply:
Looks like we have convergence on adopting the contemporary code style for tests. @arunwise would you be kind enough to finish up the PR with the reworked test suite from the gist? (Just a heads-up, it may be imperfect, as I didn’t do a full checkout of the repository.)

Reply:
@ericmjl I reworked the unit-tests to use fixtures. Most of it is based on your gist. One minor issue was that pytest has deprecated calling fixtures directly. Since, the fixtures are used as plain functions in some of the tests, I had to resort to the approach described [here](https://docs.pytest.org/en/latest/deprecations.html#id24). 

Please let me know your thoughts on the reworked test suite.